File: LeetCode/242ValidAnagram/ValidAnagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        # Sorting -> Time:O(slog(s)+tlog(t)), Space:O(1) ~ optimized, O(n) ~ worst case
        return sorted(s) == sorted(t)

        # Pythonic -> Time:O(s+t), Space:O(s+t) [worst case]
        return Counter(s) == Counter(t)

        # 2 dictionaries -> Time:O(s+t), Space:O(s+t) [worst case]
        s_d = {}
        t_d = {}
        for i in s:
            if i not in s_d:
                s_d[i] = 0
            s_d[i] += 1
        for j in t:
            if j not in t_d:
                t_d[j] = 0
            t_d[j] += 1
        return s_d == t_d

File: LeetCode/242ValidAnagram/ValidAnagram.cpp
class Solution {
public:
    bool isAnagram(string s, string t) {
        // Sorting -> Time:O(slog(s)+tlog(t)), Space:O(1) ~ optimized, O(n) ~ worst case
        sort(s.begin(), s.end());
        sort(t.begin(), t.end());
        return s == t;

        // 2 dictionaries -> Time:O(s+t), Space:O(s+t) [worst case]
        std::map<char, int> s_map;
        std::map<char, int> t_map;
        for (int i = 0; i < s.size(); i++) {
            if (s_map.find(s[i]) == s_map.end()) {
                s_map[s[i]] = 0;
            }
            s_map[s[i]] += 1;
        }
        for (int i = 0; i < t.size(); i++) {
            if (t_map.find(t[i]) == t_map.end()) {
                t_map[t[i]] = 0;
            }
            t_map[t[i]] += 1;
        }
        return s_map == t_map;
    }
};
